Experiment with partializing functions

JLockerman opened the pull request:
With this commit running

    select partialize(avg(i)) from foo;

works, as does for every other partializable aggregate postgres provides, and returns the bytea serialized form of the aggregate context. Attempting to mix partialized and non-partialized aggregates does _not_ work, but I'm not sure we care at the moment.

JLockerman committed:
With this commit running

    select partialize(avg(i)) from foo;

kinda works, and returns the bytea serialized form of the aggregate context
(I think). There is absolutely no errorchecking, so passing in an aggregate
without a serialize function will segfault.

JLockerman commented:
Example explain output for `explain (verbose) select partialize(avg(i)) from _timescaledb_internal._hyper_1_2_chunk;`
```
                                           QUERY PLAN                                           
------------------------------------------------------------------------------------------------
 Partial Aggregate  (cost=41.88..41.89 rows=1 width=32)
   Output: partialize(PARTIAL avg(i))
   ->  Seq Scan on _timescaledb_internal._hyper_1_2_chunk  (cost=0.00..35.50 rows=2550 width=4)
         Output: i
(4 rows)
```

davidkohn88 commented on src/plan_partialize.c:
```diff
@@ -0,0 +1,100 @@
+ * return type is correct.
+ */
+TSDLLEXPORT Datum
+ts_partialize(PG_FUNCTION_ARGS)
```
So, I'm thinking that we're going to need a departialize function or something similar as well.

JLockerman replied:
`departialize` is much more difficult as it requires knowing ahead of time what the internal state we're deserializing to is. It'll likely be much more invasive.

davidkohn88 reviewed (commented):
good start.

JLockerman committed:
checkpoint

JLockerman committed:
Partialize function works for all postgres supplied aggregates

We currently disallow statements with some partialized and some non-partialized
aggregates.

JLockerman commented:
**NOTE** I'm not entirely certain this is correct with `min` and `max` as postgres optimizes those specially

JLockerman committed:
Add function to turn bytea partials back to the internal state

This will call the deserialize function if it exists, otherwise it'll
use the binary input function.

JLockerman commented:
Just pushed a deserialze function which works like so
```
postgres=# select partialize(avg(i)) from _timescaledb_internal._hyper_1_2_chunk;
                                         partialize                                         
--------------------------------------------------------------------------------------------
 \x0000000100000000000000140000000200000001000000080000000000000032000000080000000000000e8d
(1 row)

postgres=# select aggregate_transition_type('avg(int)');
 aggregate_transition_type 
---------------------------
 bigint[]
(1 row)

postgres=# select aggregate_deserialize('\x0000000100000000000000140000000200000001000000080000000000000032000000080000000000000e8d', 'avg(int)', NULL::bigint[]);
WARNING:  cache reference leak: cache pg_aggregate (0), tuple 0/2 has count 1
 aggregate_deserialize 
-----------------------
 {50,3725}
(1 row)
```
   
It calls the aggregate's deserialize function if it exists, otherwise it calls the binary input function for the transition type.  
**TODO** I'm not sure yet if this is the most useful behavior, it may pay to only deserialize if it's using a builtin transition type (ie. one which would _not_ normally be turned into a `bytea`) and let postgres handle everything from there.

JLockerman commented:
also added functions
```SQL
aggregate_transition_type(REGPROCEDURE)
```
which returns the transition type of an aggregate and
```SQL
aggregate_deserialize_fn(REGPROCEDURE)
```
which returns the deserialization function the aggregate uses, if any.

JLockerman commented on src/plan_partialize.c:
```diff
@@ -0,0 +1,235 @@
+		fmgr_info(aggform->aggdeserialfn, &deserialize_finfo);
+
+		deserialized = FunctionCall1Coll(&deserialize_finfo, InvalidOid, PointerGetDatum(serialized));
+	}
```
If we decide we want postgres to handle deserialization for `internal` datatypes, we simply remove the `if` part of this conditional.
(I suspect this _will_ ultimately end up being the strategy, but we need to dig a little deeper into `Finalize Aggregate` to be sure)

JLockerman committed:
Update partialize with the correct signature
